### DISCUSSION TITLE
Using generics in generic type constraint

### DISCUSSION THREAD

**Discussion by mitkodi** [dotnet/csharplang#3208] [Language Ideas]
Sorry if it was answered somewhere, but I can't find it.
What about to have something like this:
```c#
class Entity<TKey>{}
class Repository<TEntity> where TEntity : Entity<TKey>{}
```
so we'll be able to instantiate a `Repository` specifying only the `TEntity` type:
```c#
new Repository<TEntity>()
```

**Comment by YairHalberstadt** [comment]
@mitkodi 
There are a number of languages which allow something similar to this.
For example some languages have associated types, higher kinder types, or ML style modules.

This is a complex area, and It's difficult to tell whether they would solve your problem, without having a fuller example of what your classes look like. It also may be that existing patterns in C# could solve your problem.

For example:
```csharp
class Entity<TKey> : IEntity{}
class Repository<TEntity> where TEntity : IEntity{}
```

**Comment by mitkodi** [comment]
Thank you @YairHalberstadt but I don't think it could solve it.
Here is a simplified version of what I had in mind:
```c#
public class Entity<TKey>
{
    public TKey Id { get; set; }
}

public class Person : Entity<int>
{
    ...
}

public class File : Entity<Guid>
{
    ....
}

public class Repositroy<TEntity> where TEntity : Entity<TKey>
{
    public TEntity GetById(TKey id)
    {
        ....
    }

    public TKey Insert(TEntity entity) {
    {
        ....
    }
}
```
so, it can be possible to do something like this:
```c#
    Repository<Person> people;
    Repository<File> files:
    ....
    var person = people.GetById(...);
    var file = files.GetById(...);
    ....
    person = new Person {...}
    var personId = people.Insert(person);

    file = new File {...}
    var fineId = files.Insert(file);
```
The main idea is. Let's have two entity types - Person and File and Person.

**Comment by YairHalberstadt** [comment]
Rust/Scala Associated types would allow you to do that as follows:

```csharp
public class Entity
{
    public abstract type TKey;
    public TKey Id { get; set; }
}

public class Person : Entity
{
    public override type TKey = int;
    ...
}

public class File : Entity
{
    public override type TKey = Guid;
    ....
}

public class Repository<TEntity> where TEntity : Entity
{
    public TEntity GetById(TEntity.TKey id)
    {
        ....
    }

    public TEntity.TKey Insert(TEntity entity) {
    {
        ....
    }
}
```

This can sometimes be nicer than using generics.

**Reply by bernd5** [reply]
It would work that way in C++, too. Just without `abstract TKey`.
It woule be written in C++ like:
```
using TKey = int;
```

**Reply by Blackclaws** [reply]
C++ usings are limited to the scope they are in however I'm not sure that would apply here

**Reply by bernd5** [reply]
In C++ you would [write](https://godbolt.org/z/34f14GdoE):
```C++
#include <array>
#include <type_traits>

using Guid = std::array<int, 4>;

class Entity
{
};

template<typename T> 
concept EntityConcept = requires {
    typename T::TKey;
};


class Person : Entity
{
    public:
        using TKey = int;  
};

class File : Entity
{
    public:
        using TKey = Guid;
};

template<
    EntityConcept TEntity
>
class Repository
{
    private:
    using TEntity_Key = typename TEntity::TKey;

    public:
    TEntity GetById(TEntity_Key id)
    {
    }

    TEntity_Key Insert(TEntity entity)
    {
    }
};

int main(){
    Repository<File> fileRepo;
    return 0;
}
```

**Comment by mitkodi** [comment]
But this pattern is also not supported, does it?

**Comment by YairHalberstadt** [comment]
@mitkodi 

Nope but it exists in other languages, which makes it much easier to design and evaluate.

**Comment by GalaxiaGuy** [comment]
Is this the same as this: https://github.com/dotnet/csharplang/issues/1328 ?

**Comment by NutzNK** [comment]
I would be great if we can code some thing similar like this.
But I have my own idea about pattern too.

1. 1st thing is generic type specialization.
```C#
public class GenericType<T> // This is default generic class.
{
    //Code here
}

public class GenericType<T> where T : int// This is specialized generic class of specified type
{
    //Code here
}

public class GenericType<T> where T : IComparable<T>// This is specialized generic class for interface.
{
    //Code here
}
```

2. 2nd thing is nested generic type declaration.
```C#
public Interface IEntity<TKey>
{
    TKey Id { get; set; }
}

public class Repository<TKey>
{
    TypeDef TEntity : IEntity<TKey>;
    public TEntity GetById(TKey id)
    {
        //Code here
    }

    public TKey Insert(TEntity entity) {
    {
        //Code here
    }
}
```

**Comment by YairHalberstadt** [comment]
@NutzNK 

That looks like a completely seperate proposal. Please open a seperate issue for it.

**Comment by NutzNK** [comment]
@YairHalberstadt Ok, the 1st item is difference but the 2nd item should be the same concept but different pattern.
let me check for existing issue for 1st item.

**Comment by Pyrdacor** [comment]
> Sorry if it was answered somewhere, but I can't find it.
> What about to have something like this:
> 
> ```cs
> class Entity<TKey>{}
> class Repository<TEntity> where TEntity : Entity<TKey>{}
> ```
> 
> so we'll be able to instantiate a `Repository` specifying only the `TEntity` type:
> 
> ```cs
> new Repository<TEntity>()
> ```

How would I specify `TKey` in your example? Is it determined by type inference?

**Comment by mitkodi** [comment]
Yes @Pyrdacor, the main idea is to be able to infer `TKey` from `TEntity`. At the moment we have to explicitly specify it (i.e. `class Repository<Tentity, TKey> where TEntity : Entity<TKey>`), which is redundant and will lead to compilation error if you try to declare an instance of `Repositroy<EntityType, KeyType>` with non compatible `KeyType`:

For example the following:
```c#
class Person : Entity<int>{}
...
Repository<Person, Guid> _repository;
```
will lead to *There is no implicit reference conversation from Person to Entity&lt;Guid&gt;*

**Comment by Joe4evr** [comment]
I don't see how this is much different from #997.

**Comment by mitkodi** [comment]
@Joe4evr, maybe the difference is that I mean only the type declaration.

**Comment by HaloFour** [comment]
Associated types do seem like they would be a clean way to syntactically solve the problem, however I don't see how that would lower to anything except normal generic type parameters in the CLR.  The CLR still requires that type to parameterize any members and there's only one way to convey that.  Furthermore, the raw name of the generic type is driven by the arity of generic type parameters which is why C# allows for `Repository<TEntity>` and `Repository<TEntity, TKey>` to be two distinct types.

I can't imagine that generics themselves would be overhauled in the CLR.  They're complicated enough, to the extent that they were almost never added to begin with.  I have a feeling issues like this would be better solved through improved generic type argument inference in the C# language so that both generic type arguments would not have to be specified if one argument could be derived from the other.

**Comment by Pyrdacor** [comment]
@HaloFour I think improved generic type inference in C# is exactly what was asked for here. Not changes to the CLR. The compiler should be able to inference `TKey` in the example. No extra types or changes to generic types are needed for this imho.

**Comment by HaloFour** [comment]
@Pyrdacor 

At the time of creation I can agree.  I still think that the syntax should reflect the arity of the generic type arguments, e.g. `var repo = new Repository<Person, _>();` or similar.  But I think that the generic type parameter is still necessary at declaration as that drives the contract of the type.

**Comment by Pyrdacor** [comment]
How about

```cs
class Entity<TKey>{}
class Repository<TEntity, TKey = _> where TEntity : Entity<TKey>{}
```

So TKey can be given but if not, it defaults to TKey of Entity. You can also pass TKey explicitly as _ as you mentioned. If the compiler can't inference the type then, it generates an error.

This could be generalized to: if a template type parameter is set to _, the type is inferenced and if that's not possible, a compile time error is generated.

**Comment by HaloFour** [comment]
Why would that need a change in the declaration?  If the compiler is going to be doing the inference it could try to base that on the existing generic parameter constraints which already establishes the relationship.  New syntax would require new metadata to support it and wouldn't work for existing types.

**Comment by Pyrdacor** [comment]
What do you mean by new syntax here? The discard operator in place of a type?

What if I want the user to explicitly provide a TKey that should not be inferenced from the underlying TKey then?

**Comment by HaloFour** [comment]
@Pyrdacor 

> What if I want the user to explicitly provide a TKey that should not be inferenced from the underlying TKey then?

Then you'd provide it at the point of consumption, like you do today.

```cs
// today
var repo = new Repository<Person, Guid>();

// proposed
var repo = new Repository<Person, _>(); // second parameter inferred as Guid
```

I don't think that the declaration of the class itself needs to be changed for that to be implemented.

**Comment by Pyrdacor** [comment]
No I mean what if I want to force the user of my class to provide a different type. I have to do it in the class declaration.

So in your example above I may want to deny the second possibility.

**Comment by HaloFour** [comment]
@Pyrdacor 

> So in your example above I may want to deny the second possibility.

Why would you care if someone else used generic type inference?  You can't stop them from doing it now where it's already supported.

**Comment by Pyrdacor** [comment]
True. Ok you're right.

**Comment by Blackclaws** [comment]
This seems to be the same as was proposed here: https://github.com/dotnet/roslyn/issues/5023 where the result was that it will probably never be supported as it is a breaking change.

I've also run into that exact same problem just now only to find that there is no neat solution right now. I would really like such a feature

**Reply by 333fred** [reply]
No, these are entirely unrelated. This discussion is about associated types and higher kinded types. The issue you linked is about type inference.

**Reply by Blackclaws** [reply]
I might just be confused about the actual difference between the two, but I was under the impression that at least taking the initial examples of this discussion and the issue I linked they seem to map pretty well on each other. 

The linked issue wants the generic type constraints to be involved in type inference/declaration, the initial example here seems to also do that.

Or is the difference that in one case you have the type inference on the class instead of function calls? 

In any case what I would be after is to not have to specify types that can be inferred by the context as in the example here where you have entities with a typed key and a repository that works with the entities without having to explicitly specify the key type on the repository or the functions.

**Reply by 333fred** [reply]
The first example in this discussion is of an associated type.
